hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/AbfsInputStream.java

```diff
@@ -223,16 +244,19 @@ private int readInternal(final long position, final byte[] b, final int offset,
 
       // queue read-aheads
       int numReadAheads = this.readAheadQueueDepth;
-      long nextSize;
       long nextOffset = position;
+      // First read to queue needs to be of readBufferSize and later
```

Comment:
Couple of things here:
1. The earlier code allowed bufferSize to be configurable whereas ReadAhead buffer size was fixed. And each time loop is done, read issued was always for bufferSize which can lead to gaps/holes in the readAhead range done.
2. There is no validation for 4MB as a fixed size for readAhead is optimal for all sequential reads. Having a higher readAhead range for apps like DFSIO which are guaranteed sequential and doing higher readAhead ranges in background can be performant. 
In this PR, the bug in point 1 is fixed and also a provision to configure readAhead buffer size is provided.

Reply:
What do you mean by gaps/holes in the readAhead range done here?
Have you done any experiments on this readAheadBlockSize config? If so, please share.

Reply:
If the buffer gets overwritten by config to 16MB, the readAhead buffer size will still remain to be 4MB as it was a code static. The loop done will start issuing readAheads in 16 MB buffer sizes, the request to readAhead will be:
offset=0, Length=16MB
offset=16MB, Length=32MB

But the readAhead buffer size is stuck at 4 MB. so it will read only:
offset=0 Length=4MB
offset=16MB Length=4MB

Gap being at 4MB to 16MB here.

This bug is getting fixed. Tests for all possible combinations here has been added to the tests of this PR.

Reply:
I don't think there is any bug in the current production code as such. As far as I understand the code the change is introduced becuase new config is introduced. 
Now my question is why not use readAheadBlockSize for the first call as well? The calls would be like 
offset=0 Length=4MB 
offset=4MB Length=4MB

Sorry to say this but honestly speaking, introducing so many configs is making the code complex and confusing.

Reply:
Read buffer size config being available to be modified, fixed read ahead buffer size and issuing read aheads by buffer size is the current prod behaviour and will function as the picture attached. This will need fixing.
And as for deprecating read buffer size config and only use the new read ahead buffer size config. The config has been available since GA, and hence deprecating it would not be feasible. (Also for clients who are disabling readAheads to use readahead buffer size for reads might be confusing too).
As for the number of different configs present for read, 1 and 2 configs already were present while this PR is introducing 3 and 4. So total of 4 configs.
1.	fs.azure.read.request.size
2.	fs.azure.readaheadqueue.depth
3.	fs.azure.read.alwaysReadBufferSize => For Gen1 migrating customers
4.	fs.azure.read.readahead.blocksize => Was one that needed fixing long back as there is no validation on 4 MB being the right size for all workloads. Just the way read buffer size can be modified.
 
All these changes are being added based on various customer issues and experiences that we are dealing with. Instead of spending our time in providing patches that can enable them to test various combinations, having these options over a config for their testing saves our dev time to improve the service. As you can see in the PR, the defaults introduced by these configs will retain the current prod behavior.
![ReadAheadBufferConfigExplanation](https://user-images.githubusercontent.com/13376553/100308566-b87e8580-2fce-11eb-8643-748848a69dbd.png)

Reply:
Have synced with @mukund-thakur over mail thread further to clear the understanding.